presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java

```diff
@@ -0,0 +1,174 @@
+        private Map<WindowNode.Specification, ImmutableMap.Builder<Symbol, Signature>> signatures =
+                new HashMap<>();
+
+        private boolean collectFunctions(WindowNode window)
+        {
+            WindowNode.Specification key = window.getSpecification();
+
+            // Either both maps must contain the specification, or neither.
+            checkState(functions.containsKey(key) == signatures.containsKey(key),
+                    "Merge context is in an inconsistent state.");
+
+            ImmutableMap.Builder<Symbol, FunctionCall> functionsBuilder = functions.get(key);
+            if (functionsBuilder == null) {
+                functionsBuilder = new ImmutableMap.Builder<>();
+            }
+
+            functionsBuilder.putAll(window.getWindowFunctions());
+            functions.put(key, functionsBuilder);
+
+            ImmutableMap.Builder<Symbol, Signature> signaturesBuilder = signatures.get(key);
+            if (signaturesBuilder == null) {
+                signaturesBuilder = new ImmutableMap.Builder<>();
+            }
+            signaturesBuilder.putAll(window.getSignatures());
+            return signatures.put(key, signaturesBuilder) == null;
+        }
+
+        private static class FunctionPair<F, S>
+        {
+            final Map<Symbol, FunctionCall> windowFunctions;
+            final Map<Symbol, Signature> signatures;
+
+            private FunctionPair(
+                    Map<Symbol, FunctionCall> windowFunctions,
+                    Map<Symbol, Signature> signatures)
+            {
+                this.windowFunctions = windowFunctions;
+                this.signatures = signatures;
+            }
+        }
+
+        private FunctionPair getCollectedFunctions(WindowNode window)
+        {
+            WindowNode.Specification key = window.getSpecification();
+
+            ImmutableMap.Builder<Symbol, FunctionCall> functionBuilder = functions.remove(key);
+            ImmutableMap.Builder<Symbol, Signature> signatureBuilder = signatures.remove(key);
+
+            // The only valid state here is that both maps contain the specification.
+            checkState(functionBuilder != null && signatureBuilder != null,
+                    "Merge context in inconsistent state or trying to merge the same window specification twice");
+
+            return new FunctionPair(functionBuilder.build(), signatureBuilder.build());
+        }
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<MergeIdenticalWindows.Context>
+    {
+        @Override
+        public PlanNode visitWindow(
+                WindowNode node,
+                RewriteContext<MergeIdenticalWindows.Context> context)
+        {
+            Context mergeContext = context.get();
+            boolean retainNode = mergeContext.collectFunctions(node);
+
+            // Don't merge window functions across nodes that aren't WindowNodes
+            if (node.getClass().isInstance(node.getSource())) {
```

Comment:
Why? That specifies the type in two places (parameter list and here) rather than just one (parameter list). OaOO?